<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>ANI and SNP distance matrices for specific STs</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/clipboard/clipboard.min.js"></script>
<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/quarto.js"></script>
<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/popper.min.js"></script>
<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/anchor.min.js"></script>
<link href="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ANI and SNP distance matrices for specific STs</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This tutorial will take <em>n</em> genome sequences and run algorithms to determine average nucleotide idenitities (ANI) and core genome single nucleotide polymorphisms (SNPs), visualising the distances as heatmaps in python.</p>
<p>This workflow uses <a href="https://github.com/ParBLiSS/FastANI?tab=readme-ov-file">fastANI</a> for ANI, <a href="https://github.com/tseemann/snippy">snippy</a> and <a href="https://github.com/tseemann/snp-dists">snp-dists</a> for SNP distances, and seaborn and matplotlib in python to visualise the distances as heatmaps. For any analysis in bash it uses the conda package manager so make sure you have that installed.</p>
</section>
<section id="part-1---seperating-genomes-into-st" class="level2">
<h2 class="anchored" data-anchor-id="part-1---seperating-genomes-into-st">Part 1 - Seperating genomes into ST</h2>
<p>First use cat to create a accessions.txt file</p>
<pre class="{bash}"><code>cat &gt; ST38_seqs.txt</code></pre>
<p>Type your accessions directly into the terminal, or alternately copy and paste these to follow this example:</p>
<pre class="{bash}"><code>E84
E98
E141
E158
E165
A45</code></pre>
<p>Type Ctrl + D to save and exit</p>
<p>You now have an accessions.txt containing the accessions you want to download.</p>
<p>Check the contents</p>
<pre class="{bash}"><code>cat ST38_seqs.txt</code></pre>
<p>Once you are happy with the accessions in your accessions.txt file, type the following command:</p>
<pre class="{bash}"><code>mkdir ST38_seqs 

cat ST38_seqs.txt | parallel ls all_assemblies_pilon/{}.fasta 

cat ST38_seqs.txt | parallel cp all_assemblies_pilon/{}.fasta ST38_seqs

ls ST38_seqs</code></pre>
</section>
<section id="part-2---ani-matrix" class="level2">
<h2 class="anchored" data-anchor-id="part-2---ani-matrix">Part 2 - ANI matrix</h2>
<section id="calculating-ani-with-bash" class="level3">
<h3 class="anchored" data-anchor-id="calculating-ani-with-bash">2.1 - Calculating ANI with bash</h3>
<p>We will be using <a href="https://github.com/ParBLiSS/FastANI?tab=readme-ov-file">fastANI</a> with <a href="https://anaconda.org/bioconda/fastani">bioconda</a> to determine ANI values for all of genomes compared against each other.</p>
<pre class="{bash}"><code># conda create -n ANI_SNP_dists -y python=3.8
conda activate ANI_SNP_dists
conda install bioconda::fastani</code></pre>
<p>Create a list of fasta genomes to determine ANI</p>
<pre class="{bash}"><code>ls ST38_seqs/*.fasta &gt; ST38_seqs/ST38_fastas.txt</code></pre>
<p>Check the list</p>
<pre class="{bash}"><code>cat ST38_seqs/ST38_fastas.txt</code></pre>
<p>For ANI of all vs all run fastANI with all the genomes in your list.</p>
<p>Choosing the <code>--matrix</code> flag will output a matrix, this is what we will use to plot the heatmap</p>
<pre class="{bash}"><code>fastANI --ql ST38_seqs/ST38_fastas.txt --rl ST38_seqs/ST38_fastas.txt -o ST38_seqs/ST38_ANI.tsv --matrix</code></pre>
<p>Rename the output matrix</p>
<pre class="{bash}"><code>cp ST38_seqs/ST38_ANI.tsv.matrix ST38_seqs/ST38_ANI_matrix.tsv
</code></pre>
</section>
<section id="visualising-ani-matrix-with-python" class="level3">
<h3 class="anchored" data-anchor-id="visualising-ani-matrix-with-python">2.2 - Visualising ANI matrix with python</h3>
<section id="install-libraries" class="level4">
<h4 class="anchored" data-anchor-id="install-libraries">2.2.1: Install libraries</h4>
<div id="36d24128" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install libraries as necessary </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install openpyxl </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install pandas </span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install matplotlib </span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install seaborn </span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install numpy</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#import importlib.metadata</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next load the libraries</p>
<div id="6ce60613" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.patches <span class="im">as</span> mpatches</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.colors <span class="im">import</span> to_rgb</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.gridspec <span class="im">import</span> GridSpec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="convert-triangular-matrix-to-full-square-matrix" class="level4">
<h4 class="anchored" data-anchor-id="convert-triangular-matrix-to-full-square-matrix">2.2.2: Convert triangular matrix to full square matrix</h4>
<p>Firstly we will covert the traingular matrix produced from fastANI to a full square matrix</p>
<div id="9c005d9e" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the file</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"../data/ST38_ANI_matrix.tsv"</span>, <span class="st">"r"</span>) <span class="im">as</span> f:</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> f.readlines()</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Skip the first line (header)</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> lines[<span class="dv">1</span>:]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Strip whitespace</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> [line.strip() <span class="cf">for</span> line <span class="kw">in</span> lines <span class="cf">if</span> line.strip()]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract strain names from the leftmost column</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>strain_names <span class="op">=</span> []</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>values <span class="op">=</span> []</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> lines:</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    parts <span class="op">=</span> line.split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    strain_names.append(parts[<span class="dv">0</span>])</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>    values.append([<span class="bu">float</span>(x) <span class="cf">for</span> x <span class="kw">in</span> parts[<span class="dv">1</span>:]])</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> <span class="bu">len</span>(strain_names)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>ani_matrix <span class="op">=</span> np.zeros((n, n))</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Fill the lower triangle</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(n):</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(values[i])):</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>        ani_matrix[i, j] <span class="op">=</span> values[i][j]</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>        ani_matrix[j, i] <span class="op">=</span> values[i][j]</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Fill the diagonal with 100s</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>np.fill_diagonal(ani_matrix, <span class="dv">100</span>)</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Create DataFrame</span></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>ani_df_ST38 <span class="op">=</span> pd.DataFrame(ani_matrix, index<span class="op">=</span>strain_names, columns<span class="op">=</span>strain_names)</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove ".fasta" from column names</span></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.columns <span class="op">=</span> ani_df_ST38.columns.<span class="bu">str</span>.replace(<span class="st">"ST38_seqs/"</span>, <span class="st">""</span>)</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.columns <span class="op">=</span> ani_df_ST38.index.<span class="bu">str</span>.replace(<span class="st">".fasta"</span>, <span class="st">""</span>)</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove ".fasta" from index names</span></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.index <span class="op">=</span> ani_df_ST38.index.<span class="bu">str</span>.replace(<span class="st">"ST38_seqs/"</span>, <span class="st">""</span>)</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.index <span class="op">=</span> ani_df_ST38.index.<span class="bu">str</span>.replace(<span class="st">".fasta"</span>, <span class="st">""</span>)</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Display to confirm</span></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ani_df_ST38.head())</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove ".fasta" from column names</span></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.columns <span class="op">=</span> ani_df_ST38.columns.<span class="bu">str</span>.replace(<span class="st">"ST38_seqs/"</span>, <span class="st">""</span>)</span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.columns <span class="op">=</span> ani_df_ST38.index.<span class="bu">str</span>.replace(<span class="st">".fasta"</span>, <span class="st">""</span>)</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Display to confirm</span></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ani_df_ST38.head())</span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>ani_df_ST38.to_csv(<span class="st">"../tbls/ST38_ANI_matrix_indonesia_trycycle.csv"</span>,  <span class="co"># path (and name) of the file to write</span></span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>    sep<span class="op">=</span><span class="st">","</span>,                <span class="co"># delimiter ("," by default)</span></span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>    index<span class="op">=</span><span class="va">False</span>,            <span class="co"># don’t write row numbers (like R’s row.names = FALSE)</span></span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>    header<span class="op">=</span><span class="va">True</span>,            <span class="co"># write out column names</span></span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>    encoding<span class="op">=</span><span class="st">"utf-8"</span>        <span class="co"># file encoding</span></span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         ST38_seqs/211_S1  ST38_seqs/222_S11  ST38_seqs/A1_S1  \
211_S1         100.000000          99.947029        99.982414   
222_S11         99.947029         100.000000        99.970627   
A1_S1           99.982414          99.970627       100.000000   
B4_S10          99.440842          99.426079        99.460754   
D5_S23          99.473862          99.444626        99.477219   

         ST38_seqs/B4_S10  ST38_seqs/D5_S23  ST38_seqs/E113  ST38_seqs/E117  \
211_S1          99.440842         99.473862       99.960251       99.966888   
222_S11         99.426079         99.444626       99.973358       99.960297   
A1_S1           99.460754         99.477219       99.963959       99.977356   
B4_S10         100.000000         99.889023       99.462616       99.440575   
D5_S23          99.889023        100.000000       99.476875       99.442474   

         ST38_seqs/E136  ST38_seqs/E153  ST38_seqs/E177  ST38_seqs/E79  \
211_S1        99.403992       99.486908       99.410881      99.724228   
222_S11       99.357895       99.453064       99.361801      99.649956   
A1_S1         99.388702       99.448425       99.394821      99.730743   
B4_S10        99.879967       99.367264       99.894257      99.411072   
D5_S23        99.845367       99.357643       99.913948      99.360580   

         ST38_seqs/H200  
211_S1        99.985268  
222_S11       99.977982  
A1_S1         99.986038  
B4_S10        99.453735  
D5_S23        99.467850  
             211_S1     222_S11       A1_S1      B4_S10      D5_S23  \
211_S1   100.000000   99.947029   99.982414   99.440842   99.473862   
222_S11   99.947029  100.000000   99.970627   99.426079   99.444626   
A1_S1     99.982414   99.970627  100.000000   99.460754   99.477219   
B4_S10    99.440842   99.426079   99.460754  100.000000   99.889023   
D5_S23    99.473862   99.444626   99.477219   99.889023  100.000000   

              E113       E117       E136       E153       E177        E79  \
211_S1   99.960251  99.966888  99.403992  99.486908  99.410881  99.724228   
222_S11  99.973358  99.960297  99.357895  99.453064  99.361801  99.649956   
A1_S1    99.963959  99.977356  99.388702  99.448425  99.394821  99.730743   
B4_S10   99.462616  99.440575  99.879967  99.367264  99.894257  99.411072   
D5_S23   99.476875  99.442474  99.845367  99.357643  99.913948  99.360580   

              H200  
211_S1   99.985268  
222_S11  99.977982  
A1_S1    99.986038  
B4_S10   99.453735  
D5_S23   99.467850  </code></pre>
</div>
</div>
</section>
<section id="write-a-function-to-create-the-ani-heatmaps" class="level4">
<h4 class="anchored" data-anchor-id="write-a-function-to-create-the-ani-heatmaps">2.2.3: Write a function to create the ANI heatmaps</h4>
<p>Next we write a function which masks half the dataset to create a triangular heatmap, with the added functionality of rotating the heatmap</p>
<div id="c5acdf2c" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_ani_heatmap(df, title<span class="op">=</span><span class="st">"ANI Heatmap"</span>, rotation<span class="op">=</span><span class="dv">0</span>, lower_legend<span class="op">=</span><span class="dv">95</span>, upper_legend<span class="op">=</span><span class="dv">100</span>):</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make a copy to avoid modifying the original</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    df_plot <span class="op">=</span> df.copy()</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Apply rotation first</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span>:</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.transpose()</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>].transpose()</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.zeros_like(df_plot, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    mask[np.triu_indices_from(mask, k<span class="op">=</span><span class="dv">0</span>)] <span class="op">=</span> <span class="va">True</span>  <span class="co"># k=1 excludes the diagonal</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span> <span class="kw">or</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.transpose(mask)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.flip(mask)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">10</span>))</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>        df_plot, </span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>        fmt<span class="op">=</span><span class="st">".2f"</span>, </span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>        mask<span class="op">=</span>mask, </span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">"coolwarm"</span>,</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>        vmin<span class="op">=</span>lower_legend,</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>        vmax<span class="op">=</span>upper_legend,</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>        annot_kws<span class="op">=</span>{<span class="st">"size"</span>: <span class="dv">8</span>},</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>        xticklabels<span class="op">=</span>df_plot.columns, </span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>        yticklabels<span class="op">=</span>df_plot.index, </span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>        cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: <span class="st">"ANI (%)"</span>}</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span><span class="dv">10</span>, rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>)</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-the-ani-heatmap" class="level4">
<h4 class="anchored" data-anchor-id="create-the-ani-heatmap">2.2.4: Create the ANI heatmap</h4>
<p>We will produce an ANI matrix for <em>E. coli</em> species types <span class="citation" data-cites="Ec-ANI">@Ec-ANI</span></p>
<div id="cell-Ec-ANI" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the function to create the heatmap</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>plot_ani_heatmap(df<span class="op">=</span>ani_df_ST38, title<span class="op">=</span><span class="st">"ST38 ANI Heatmap"</span>, rotation <span class="op">=</span><span class="dv">90</span>, lower_legend<span class="op">=</span><span class="fl">99.5</span>, upper_legend<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="ec-ani" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/ec-ani-output-1.png" width="1085" height="949" class="figure-img"></p>
<figcaption>ST38 ANI heatmap</figcaption>
</figure>
</div>
</div>
</div>
<p>Rodriguez et. al (2024) analysed 18,123 genomes to determine where the thresholds lay which distinguished certain taxonomic ranks:</p>
<p>same species - 95% same sequence type - 99.5% same strain - 99.9%</p>
</section>
<section id="write-a-function-to-create-the-ani-heatmaps-and-save" class="level4">
<h4 class="anchored" data-anchor-id="write-a-function-to-create-the-ani-heatmaps-and-save">2.2.5: Write a function to create the ANI heatmaps and save</h4>
<p>Next we write a function which masks half the dataset to create a triangular heatmap, with the added functionality of rotating the heatmap</p>
<div id="8c7dbcd7" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> save_ani_heatmap(df, title<span class="op">=</span><span class="st">"ANI Heatmap"</span>, rotation<span class="op">=</span><span class="dv">0</span>, lower_legend<span class="op">=</span><span class="dv">95</span>, upper_legend<span class="op">=</span><span class="dv">100</span>, output_file_png <span class="op">=</span> <span class="st">"ani_heatmap_from_table_A_Z.png"</span>, output_file_svg <span class="op">=</span> <span class="st">"ani_heatmap_from_table_A_Z.svg"</span>):</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make a copy to avoid modifying the original</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    df_plot <span class="op">=</span> df.copy()</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Apply rotation first</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span>:</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.transpose()</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>].transpose()</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.zeros_like(df_plot, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    mask[np.triu_indices_from(mask, k<span class="op">=</span><span class="dv">0</span>)] <span class="op">=</span> <span class="va">True</span>  <span class="co"># k=0 excludes the diagonal</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span> <span class="kw">or</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.transpose(mask)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.flip(mask)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">10</span>))</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>        df_plot, </span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>        fmt<span class="op">=</span><span class="st">".2f"</span>, </span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>        mask<span class="op">=</span>mask, </span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">"coolwarm"</span>,</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>        vmin<span class="op">=</span>lower_legend,</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>        vmax<span class="op">=</span>upper_legend,</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>        annot_kws<span class="op">=</span>{<span class="st">"size"</span>: <span class="dv">8</span>},</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>        xticklabels<span class="op">=</span>df_plot.columns, </span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>        yticklabels<span class="op">=</span>df_plot.index, </span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>        cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: <span class="st">"ANI (%)"</span>}</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span><span class="dv">10</span>, rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>)</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_file_png, dpi<span class="op">=</span><span class="dv">300</span>)  <span class="co"># Save the heatmap</span></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_file_svg, <span class="bu">format</span><span class="op">=</span><span class="st">"svg"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Heatmap saved to </span><span class="sc">{</span>output_file_png<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Heatmap saved to </span><span class="sc">{</span>output_file_svg<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="save-the-ani-heatmap" class="level4">
<h4 class="anchored" data-anchor-id="save-the-ani-heatmap">2.2.4: Save the ANI heatmap</h4>
<div id="00f820ac" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ST38</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>save_ani_heatmap(df<span class="op">=</span>ani_df_ST38, title<span class="op">=</span><span class="st">"ST38 ANI Heatmap"</span>, output_file_png<span class="op">=</span><span class="st">"../imgs/ani_heatmap_for_ST38_trycycle.png"</span>, output_file_svg<span class="op">=</span><span class="st">"../imgs/ani_heatmap_for_ST38_trycycle.svg"</span>, rotation<span class="op">=</span><span class="dv">90</span>, lower_legend<span class="op">=</span><span class="fl">99.5</span>, upper_legend<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Heatmap saved to ../imgs/ani_heatmap_for_ST38_trycycle.png
Heatmap saved to ../imgs/ani_heatmap_for_ST38_trycycle.svg</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/cell-8-output-2.png" width="1085" height="949" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Determine the sample with the highest associations across all samples i.e.&nbsp;the most interconnected sample - we will use this for the reference for snippy</p>
<div id="c8bb2a2e" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace diagonal with NaN so self-comparisons don't inflate the result</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>ani_df_ST38_no_diag <span class="op">=</span> ani_df_ST38.copy()</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> ani_df_ST38_no_diag.index:</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    ani_df_ST38_no_diag.loc[i, i] <span class="op">=</span> <span class="va">None</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the average ANI to other samples for each sample</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>mean_ani_ST38 <span class="op">=</span> ani_df_ST38_no_diag.mean(axis<span class="op">=</span><span class="dv">1</span>)  <span class="co"># row-wise mean</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort to find the sample with the highest connectivity</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>most_connected_ST38 <span class="op">=</span> mean_ani_ST38.sort_values(ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Show top 5 most interconnected samples</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Most interconnected samples in ST38 based on mean ANI:"</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(most_connected_ST38.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Most interconnected samples in ST38 based on mean ANI:
H200      99.711721
211_S1    99.707506
A1_S1     99.707369
E113      99.702396
E117      99.701797
dtype: float64</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="part-3---snp-distance-matrix" class="level2">
<h2 class="anchored" data-anchor-id="part-3---snp-distance-matrix">Part 3 - SNP distance matrix</h2>
<section id="calculating-snp-distances-with-bash" class="level3">
<h3 class="anchored" data-anchor-id="calculating-snp-distances-with-bash">3.1 - Calculating SNP distances with bash</h3>
<p>We will determine SNP distances with <a href="https://github.com/tseemann/snippy">snippy</a> and <a href="https://github.com/tseemann/snp-dists">snp-dists</a> with 9A-1-1 as reference</p>
<p>We will activate the same conda environment used previously in <a href="#calculating-ani-with-bash">section 2.1</a>.</p>
<p>But here we will add more programs:</p>
<ul>
<li><a href="https://github.com/tseemann/snippy">snippy</a></li>
<li><a href="https://github.com/tseemann/snp-dists">snp-dists</a></li>
<li><a href="https://www.gnu.org/software/parallel/">parallel</a></li>
</ul>
<section id="download-software" class="level4">
<h4 class="anchored" data-anchor-id="download-software">3.1.1: Download software</h4>
<pre class="{bash}"><code>conda activate ANI_SNP_dists
conda install -c conda-forge -c bioconda -c defaults snippy
conda install -c bioconda -c conda-forge snp-dists
conda install bioconda::parallel</code></pre>
<div id="19e5a981" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show top 5 most interconnected samples</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Most interconnected samples in ST38 based on mean ANI:"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(most_connected_ST38.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Most interconnected samples in ST38 based on mean ANI:
H200      99.711721
211_S1    99.707506
A1_S1     99.707369
E113      99.702396
E117      99.701797
dtype: float64</code></pre>
</div>
</div>
</section>
<section id="run-snippy" class="level4">
<h4 class="anchored" data-anchor-id="run-snippy">3.1.2: Run Snippy</h4>
<p>Use <a href="https://github.com/tseemann/snippy"><code>snippy</code></a> to generate all SNPs.</p>
<p>Set reference file for SNP calculations.</p>
<pre class="{bash}"><code># Change this time 
REF=ST38_seqs/H200.fasta</code></pre>
<p>Use <code>sed</code> to remove <code>.fastq</code> from <code>.txt</code> file</p>
<pre class="{bash}"><code>ls ST38_seqs/*.fasta &gt; ST38_seqs/ST38_fastas.txt
sed 's|ST38_seqs/||g' ST38_seqs/ST38_fastas.txt &gt; ST38_seqs/genome_names_1.txt
sed -e 's/\.fasta.*//' ST38_seqs/genome_names_1.txt &gt; ST38_seqs/genome_names.txt</code></pre>
<p>Check the new <code>.txt</code> file containing list of genome names</p>
<pre class="{bash}"><code>cat ST38_seqs/genome_names.txt

cat ST38_seqs/genome_names.txt | parallel ls ST38_seqs/{}.fasta</code></pre>
<p>The we use <a href="https://www.gnu.org/software/parallel/">parallel</a> on our list of genomes to run snippy:</p>
<pre class="{bash}"><code>cat ST38_seqs/genome_names.txt | parallel snippy --report --outdir ST38_seqs/{}_snps --ref $REF --ctgs ST38_seqs/{}.fasta</code></pre>
<p>This produces several files:</p>
</section>
<section id="run-snippy-core" class="level4">
<h4 class="anchored" data-anchor-id="run-snippy-core">3.1.3: Run Snippy-core</h4>
<p>Use <code>snippy-core</code> from <a href="https://github.com/tseemann/snippy">snippy</a> to generate core SNPs</p>
<pre class="{bash}"><code>snippy-core --ref $REF --prefix core ST38_seqs/*_snps
# move files 
mv *core* ST38_seqs/</code></pre>
<p>This produces several files:</p>
<ul>
<li><code>core.full.aln</code>: The full core genome alignment in FASTA format.</li>
<li><code>core.aln</code>: The core SNP alignment in FASTA format (only variable sites).</li>
<li><code>core.tab</code>: A table summarizing the SNP differences.</li>
</ul>
</section>
<section id="generate-a-pairwise-snp-distance-matrix" class="level4">
<h4 class="anchored" data-anchor-id="generate-a-pairwise-snp-distance-matrix">3.1.4: Generate a Pairwise SNP Distance Matrix</h4>
<p>Once you have the core SNP alignment (<code>core.aln</code>), use <a href="https://github.com/tseemann/snp-dists"><code>snp-dists</code></a> to calculate pairwise SNP distances.</p>
<pre class="{bash}"><code>snp-dists ST38_seqs/core.aln &gt; ST38_seqs/ST38_snp_matrix.tsv</code></pre>
<p>This will generate a pairwise SNP distance matrix (<code>snp_matrix.tsv</code>) where:</p>
<ul>
<li>Rows and columns correspond to isolates.</li>
<li>The values represent the number of SNP differences between isolates</li>
</ul>
</section>
</section>
<section id="visualising-snp-distance-matrix-with-python" class="level3">
<h3 class="anchored" data-anchor-id="visualising-snp-distance-matrix-with-python">3.2 - Visualising SNP distance matrix with python</h3>
<p>Make sure you have all the required libraries installed, if you need to install them see <a href="#install-libraries">section 2.2.1</a></p>
<section id="read-and-clean-the-data" class="level4">
<h4 class="anchored" data-anchor-id="read-and-clean-the-data">3.2.1: Read and clean the data</h4>
<div id="5a0bcc96" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the SNP matrix with first column as row index</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>snp_df_ST38 <span class="op">=</span> pd.read_csv(<span class="st">"../data/ST38_snp_matrix.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove reference file</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>snp_df_ST38 <span class="op">=</span> snp_df_ST38.drop(<span class="st">"Reference"</span>, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>snp_df_ST38 <span class="op">=</span> snp_df_ST38.drop(<span class="st">"Reference"</span>, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove "_snps" from column names</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>snp_df_ST38.columns <span class="op">=</span> snp_df_ST38.columns.<span class="bu">str</span>.replace(<span class="st">"_snps"</span>, <span class="st">""</span>)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove "_snps" from index names</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>snp_df_ST38.index <span class="op">=</span> snp_df_ST38.index.<span class="bu">str</span>.replace(<span class="st">"_snps"</span>, <span class="st">""</span>)</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Verify it loaded correctly</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(snp_df_ST38.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 211_S1  222_S11  A1_S1  B4_S10  D5_S23   E113   E117   E136  \
snp-dists 0.8.2                                                                
211_S1                0       35     34   13076   13091     36     33  13102   
222_S11              35        0     45   13059   13074     13     44  13085   
A1_S1                34       45      0   13086   13099     46     43  13112   
B4_S10            13076    13059  13086       0     167  13060  13085    180   
D5_S23            13091    13074  13099     167       0  13075  13100    187   

                  E153   E177    E79   H200  
snp-dists 0.8.2                              
211_S1            8969  11859   2454     33  
222_S11           8952  11844   2437     44  
A1_S1             8979  11867   2464      1  
B4_S10           15624   1430  13149  13085  
D5_S23           15637   1437  13164  13100  </code></pre>
</div>
</div>
</section>
<section id="create-a-function-which-makes-a-snp-distance-heatmap" class="level4">
<h4 class="anchored" data-anchor-id="create-a-function-which-makes-a-snp-distance-heatmap">3.2.2: Create a function which makes a SNP distance heatmap</h4>
<div id="ca0052f9" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_snp_heatmap(df, title<span class="op">=</span><span class="st">"SNP Heatmap"</span>, rotation<span class="op">=</span><span class="dv">0</span>, lower_legend<span class="op">=</span><span class="dv">95</span>, upper_legend<span class="op">=</span><span class="dv">100</span>):</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make a copy to avoid modifying the original</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    df_plot <span class="op">=</span> df.copy()</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Apply rotation first</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span>:</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.transpose()</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>].transpose()</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.zeros_like(df_plot, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>    mask[np.triu_indices_from(mask)] <span class="op">=</span> <span class="va">True</span>  <span class="co"># k=1 excludes the diagonal</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span> <span class="kw">or</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.transpose(mask)</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.flip(mask)</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">10</span>))</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>        df_plot, </span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>        fmt<span class="op">=</span><span class="st">".2f"</span>, </span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>        mask<span class="op">=</span>mask, </span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">"RdBu"</span>,</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>        vmin<span class="op">=</span>lower_legend,</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>        vmax<span class="op">=</span>upper_legend,</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>        annot_kws<span class="op">=</span>{<span class="st">"size"</span>: <span class="dv">8</span>},</span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>        xticklabels<span class="op">=</span>df_plot.columns, </span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>        yticklabels<span class="op">=</span>df_plot.index, </span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a>        cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: <span class="st">"SNP count"</span>}</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span><span class="dv">10</span>, rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>)</span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="create-the-snp-distance-heatmaps" class="level3">
<h3 class="anchored" data-anchor-id="create-the-snp-distance-heatmaps">3.2.3: Create the SNP distance heatmaps</h3>
<p>Next we will produce an SNP distance matrix for <em>E. coli</em> species types <span class="citation" data-cites="Ec-SNP">@Ec-SNP</span></p>
<div id="cell-Ec-SNP" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the function to create the heatmap</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>fig_title <span class="op">=</span> <span class="st">"ST38 SNP distance heatmap"</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>create_snp_heatmap(title<span class="op">=</span>fig_title, df <span class="op">=</span> snp_df_ST38, rotation<span class="op">=</span><span class="dv">90</span>, lower_legend<span class="op">=</span><span class="dv">0</span>, upper_legend<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="ec-snp" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/ec-snp-output-1.png" width="1072" height="949" class="figure-img"></p>
<figcaption>E. coli SNP distance heatmap</figcaption>
</figure>
</div>
</div>
</div>
<section id="write-a-function-to-create-the-snp-heatmaps-and-save" class="level4">
<h4 class="anchored" data-anchor-id="write-a-function-to-create-the-snp-heatmaps-and-save">2.2.5: Write a function to create the SNP heatmaps and save</h4>
<p>Next we write a function which masks half the dataset to create a triangular heatmap, with the added functionality of rotating the heatmap</p>
<div id="45417e3f" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> save_snp_heatmap(df, title<span class="op">=</span><span class="st">"SNP Heatmap"</span>, rotation<span class="op">=</span><span class="dv">0</span>, lower_legend<span class="op">=</span><span class="dv">95</span>, upper_legend<span class="op">=</span><span class="dv">100</span>, output_file_png <span class="op">=</span> <span class="st">"snp_heatmap_from_table_A_Z.png"</span>, output_file_svg <span class="op">=</span> <span class="st">"snp_heatmap_from_table_A_Z.svg"</span>):</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make a copy to avoid modifying the original</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    df_plot <span class="op">=</span> df.copy()</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Apply rotation first</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span>:</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.transpose()</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>        df_plot <span class="op">=</span> df_plot.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>].transpose()</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.zeros_like(df_plot, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    mask[np.triu_indices_from(mask, k<span class="op">=</span><span class="dv">0</span>)] <span class="op">=</span> <span class="va">True</span>  <span class="co"># k=0 excludes the diagonal</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation <span class="op">==</span> <span class="dv">90</span> <span class="kw">or</span> rotation <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.transpose(mask)</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> rotation <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>        mask <span class="op">=</span> np.flip(mask)</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">10</span>))</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>        df_plot, </span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>        fmt<span class="op">=</span><span class="st">".0f"</span>, </span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>        mask<span class="op">=</span>mask, </span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">"RdBu"</span>,</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>        vmin<span class="op">=</span>lower_legend,</span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>        vmax<span class="op">=</span>upper_legend,</span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>        annot_kws<span class="op">=</span>{<span class="st">"size"</span>: <span class="dv">8</span>},</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>        xticklabels<span class="op">=</span>df_plot.columns, </span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>        yticklabels<span class="op">=</span>df_plot.index, </span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>        cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: <span class="st">"Number of SNPs"</span>}</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span><span class="dv">10</span>, rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>)</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_file_png, dpi<span class="op">=</span><span class="dv">300</span>)  <span class="co"># Save the heatmap</span></span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_file_svg, <span class="bu">format</span><span class="op">=</span><span class="st">"svg"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Heatmap saved to </span><span class="sc">{</span>output_file_png<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Heatmap saved to </span><span class="sc">{</span>output_file_svg<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="save-the-snp-heatmap" class="level4">
<h4 class="anchored" data-anchor-id="save-the-snp-heatmap">2.2.4: Save the SNP heatmap</h4>
<div id="3fa4ce16" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ST38</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>save_snp_heatmap(df<span class="op">=</span>snp_df_ST38, title<span class="op">=</span><span class="st">"ST38 SNP Heatmap"</span>, output_file_png<span class="op">=</span><span class="st">"../imgs/snp_heatmap_for_ST38_trycycle.png"</span>, output_file_svg<span class="op">=</span><span class="st">"../imgs/snp_heatmap_for_ST38_trycycle.svg"</span>, rotation<span class="op">=</span><span class="dv">90</span>, lower_legend<span class="op">=</span><span class="dv">0</span>, upper_legend<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Heatmap saved to ../imgs/snp_heatmap_for_ST38_trycycle.png
Heatmap saved to ../imgs/snp_heatmap_for_ST38_trycycle.svg</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/cell-15-output-2.png" width="1072" height="949" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="part-4---combined-ani-and-snp-dist-matrix" class="level2">
<h2 class="anchored" data-anchor-id="part-4---combined-ani-and-snp-dist-matrix">Part 4 - Combined ANI and SNP dist matrix</h2>
<section id="combined-ani-and-snp-values" class="level3">
<h3 class="anchored" data-anchor-id="combined-ani-and-snp-values">4.1 - Combined ANI and SNP values</h3>
<div id="b5ae46b5" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> save_dual_annot_heatmap(snp_df,</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>                            ani_df,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                            color_by<span class="op">=</span><span class="st">"ani"</span>,               <span class="co"># "ani" or "snp"</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>                            title<span class="op">=</span><span class="st">"Dual‐Annotated Heatmap"</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>                            rotation<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>                            lower_legend<span class="op">=</span><span class="va">None</span>,            <span class="co"># if None, auto from data</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                            upper_legend<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                            output_png<span class="op">=</span><span class="st">"dual_heatmap.png"</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>                            output_svg<span class="op">=</span><span class="st">"dual_heatmap.svg"</span>):</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1) Quick sanity checks</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> snp_df.shape <span class="op">==</span> ani_df.shape, <span class="st">"Shapes must match"</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> <span class="bu">all</span>(snp_df.index <span class="op">==</span> ani_df.index) <span class="kw">and</span> <span class="bu">all</span>(snp_df.columns <span class="op">==</span> ani_df.columns)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 2) Pick the matrix that drives the color</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> color_by <span class="op">==</span> <span class="st">"ani"</span>:</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>        cmap_df <span class="op">=</span> ani_df</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>        fmt <span class="op">=</span> <span class="st">"</span><span class="sc">{:.2f}</span><span class="st">%"</span></span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a>        cbar_label <span class="op">=</span> <span class="st">"ANI (%)"</span></span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> color_by <span class="op">==</span> <span class="st">"snp"</span>:</span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a>        cmap_df <span class="op">=</span> snp_df</span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a>        fmt <span class="op">=</span> <span class="st">"</span><span class="sc">{:.0f}</span><span class="st">"</span></span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a>        cbar_label <span class="op">=</span> <span class="st">"SNP distance"</span></span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a>        <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="st">"color_by must be 'ani' or 'snp'"</span>)</span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-26"><a href="#cb37-26" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 3) Optionally rotate</span></span>
<span id="cb37-27"><a href="#cb37-27" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> _rotate(df, rot):</span>
<span id="cb37-28"><a href="#cb37-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> rot <span class="op">==</span> <span class="dv">90</span>:</span>
<span id="cb37-29"><a href="#cb37-29" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> df.T</span>
<span id="cb37-30"><a href="#cb37-30" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> rot <span class="op">==</span> <span class="dv">180</span>:</span>
<span id="cb37-31"><a href="#cb37-31" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> df.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb37-32"><a href="#cb37-32" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> rot <span class="op">==</span> <span class="dv">270</span>:</span>
<span id="cb37-33"><a href="#cb37-33" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> df.iloc[::<span class="op">-</span><span class="dv">1</span>, ::<span class="op">-</span><span class="dv">1</span>].T</span>
<span id="cb37-34"><a href="#cb37-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb37-35"><a href="#cb37-35" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> df</span>
<span id="cb37-36"><a href="#cb37-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-37"><a href="#cb37-37" aria-hidden="true" tabindex="-1"></a>    cmap_df   <span class="op">=</span> _rotate(cmap_df, rotation)</span>
<span id="cb37-38"><a href="#cb37-38" aria-hidden="true" tabindex="-1"></a>    ani_df_r  <span class="op">=</span> _rotate(ani_df, rotation)</span>
<span id="cb37-39"><a href="#cb37-39" aria-hidden="true" tabindex="-1"></a>    snp_df_r  <span class="op">=</span> _rotate(snp_df, rotation)</span>
<span id="cb37-40"><a href="#cb37-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-41"><a href="#cb37-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 4) Mask upper triangle (optional; remove if you want full matrix)</span></span>
<span id="cb37-42"><a href="#cb37-42" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.zeros_like(cmap_df, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb37-43"><a href="#cb37-43" aria-hidden="true" tabindex="-1"></a>    mask[np.triu_indices_from(mask, k<span class="op">=</span><span class="dv">0</span>)] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb37-44"><a href="#cb37-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-45"><a href="#cb37-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 5) Determine color‐scale bounds</span></span>
<span id="cb37-46"><a href="#cb37-46" aria-hidden="true" tabindex="-1"></a>    vmin <span class="op">=</span> lower_legend <span class="cf">if</span> lower_legend <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> np.nanmin(cmap_df.values)</span>
<span id="cb37-47"><a href="#cb37-47" aria-hidden="true" tabindex="-1"></a>    vmax <span class="op">=</span> upper_legend <span class="cf">if</span> upper_legend <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> np.nanmax(cmap_df.values)</span>
<span id="cb37-48"><a href="#cb37-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-49"><a href="#cb37-49" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 6) Plot!</span></span>
<span id="cb37-50"><a href="#cb37-50" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">10</span>))</span>
<span id="cb37-51"><a href="#cb37-51" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb37-52"><a href="#cb37-52" aria-hidden="true" tabindex="-1"></a>        cmap_df,</span>
<span id="cb37-53"><a href="#cb37-53" aria-hidden="true" tabindex="-1"></a>        mask<span class="op">=</span>mask,</span>
<span id="cb37-54"><a href="#cb37-54" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">"coolwarm"</span> <span class="cf">if</span> color_by<span class="op">==</span><span class="st">"ani"</span> <span class="cf">else</span> <span class="st">"RdBu_r"</span>,</span>
<span id="cb37-55"><a href="#cb37-55" aria-hidden="true" tabindex="-1"></a>        vmin<span class="op">=</span>vmin,</span>
<span id="cb37-56"><a href="#cb37-56" aria-hidden="true" tabindex="-1"></a>        vmax<span class="op">=</span>vmax,</span>
<span id="cb37-57"><a href="#cb37-57" aria-hidden="true" tabindex="-1"></a>        cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: cbar_label},</span>
<span id="cb37-58"><a href="#cb37-58" aria-hidden="true" tabindex="-1"></a>        xticklabels<span class="op">=</span>cmap_df.columns,</span>
<span id="cb37-59"><a href="#cb37-59" aria-hidden="true" tabindex="-1"></a>        yticklabels<span class="op">=</span>cmap_df.index,</span>
<span id="cb37-60"><a href="#cb37-60" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">False</span></span>
<span id="cb37-61"><a href="#cb37-61" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb37-62"><a href="#cb37-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-63"><a href="#cb37-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 7) Overlay both ANI and SNP text</span></span>
<span id="cb37-64"><a href="#cb37-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, row <span class="kw">in</span> <span class="bu">enumerate</span>(cmap_df.index):</span>
<span id="cb37-65"><a href="#cb37-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j, col <span class="kw">in</span> <span class="bu">enumerate</span>(cmap_df.columns):</span>
<span id="cb37-66"><a href="#cb37-66" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> mask[i, j]:</span>
<span id="cb37-67"><a href="#cb37-67" aria-hidden="true" tabindex="-1"></a>                <span class="cf">continue</span></span>
<span id="cb37-68"><a href="#cb37-68" aria-hidden="true" tabindex="-1"></a>            ani_val <span class="op">=</span> ani_df_r.iloc[i, j]</span>
<span id="cb37-69"><a href="#cb37-69" aria-hidden="true" tabindex="-1"></a>            snp_val <span class="op">=</span> snp_df_r.iloc[i, j]</span>
<span id="cb37-70"><a href="#cb37-70" aria-hidden="true" tabindex="-1"></a>            <span class="co"># two‐line text: ANI% on top, SNP below</span></span>
<span id="cb37-71"><a href="#cb37-71" aria-hidden="true" tabindex="-1"></a>            txt <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>ani_val<span class="sc">:.1f}</span><span class="ss">%</span><span class="ch">\n</span><span class="sc">{</span>snp_val<span class="sc">:.0f}</span><span class="ss">"</span></span>
<span id="cb37-72"><a href="#cb37-72" aria-hidden="true" tabindex="-1"></a>            plt.text(j <span class="op">+</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>, txt,</span>
<span id="cb37-73"><a href="#cb37-73" aria-hidden="true" tabindex="-1"></a>                     ha<span class="op">=</span><span class="st">"center"</span>, va<span class="op">=</span><span class="st">"center"</span>,</span>
<span id="cb37-74"><a href="#cb37-74" aria-hidden="true" tabindex="-1"></a>                     fontsize<span class="op">=</span><span class="dv">8</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb37-75"><a href="#cb37-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-76"><a href="#cb37-76" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 8) Finish touches</span></span>
<span id="cb37-77"><a href="#cb37-77" aria-hidden="true" tabindex="-1"></a>    plt.title(title, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb37-78"><a href="#cb37-78" aria-hidden="true" tabindex="-1"></a>    plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb37-79"><a href="#cb37-79" aria-hidden="true" tabindex="-1"></a>    plt.yticks(rotation<span class="op">=</span><span class="dv">0</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb37-80"><a href="#cb37-80" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb37-81"><a href="#cb37-81" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_png, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb37-82"><a href="#cb37-82" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_svg, <span class="bu">format</span><span class="op">=</span><span class="st">"svg"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb37-83"><a href="#cb37-83" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb37-84"><a href="#cb37-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-85"><a href="#cb37-85" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Saved heatmap to </span><span class="sc">{</span>output_png<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>output_svg<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-86"><a href="#cb37-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-87"><a href="#cb37-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-88"><a href="#cb37-88" aria-hidden="true" tabindex="-1"></a><span class="co"># Color by ANI, annotate with both ANI+SNP:</span></span>
<span id="cb37-89"><a href="#cb37-89" aria-hidden="true" tabindex="-1"></a>save_dual_annot_heatmap(</span>
<span id="cb37-90"><a href="#cb37-90" aria-hidden="true" tabindex="-1"></a>    snp_df<span class="op">=</span>snp_df_ST38,</span>
<span id="cb37-91"><a href="#cb37-91" aria-hidden="true" tabindex="-1"></a>    ani_df<span class="op">=</span>ani_df_ST38,</span>
<span id="cb37-92"><a href="#cb37-92" aria-hidden="true" tabindex="-1"></a>    color_by<span class="op">=</span><span class="st">"ani"</span>,</span>
<span id="cb37-93"><a href="#cb37-93" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"ST38: ANI‐color &amp; SNP‐annot"</span>,</span>
<span id="cb37-94"><a href="#cb37-94" aria-hidden="true" tabindex="-1"></a>    rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb37-95"><a href="#cb37-95" aria-hidden="true" tabindex="-1"></a>    lower_legend<span class="op">=</span><span class="fl">99.5</span>,   <span class="co"># for the ANI colorbar</span></span>
<span id="cb37-96"><a href="#cb37-96" aria-hidden="true" tabindex="-1"></a>    upper_legend<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb37-97"><a href="#cb37-97" aria-hidden="true" tabindex="-1"></a>    output_png<span class="op">=</span><span class="st">"../imgs/ani_and_snp_heatmap_colored_by_ani_ST38.png"</span>,</span>
<span id="cb37-98"><a href="#cb37-98" aria-hidden="true" tabindex="-1"></a>    output_svg<span class="op">=</span><span class="st">"../imgs/ani_and_snp_heatmap_colored_by_ani_ST38.svg"</span></span>
<span id="cb37-99"><a href="#cb37-99" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb37-100"><a href="#cb37-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-101"><a href="#cb37-101" aria-hidden="true" tabindex="-1"></a><span class="co"># Or, color by SNP distance, annotate with both:</span></span>
<span id="cb37-102"><a href="#cb37-102" aria-hidden="true" tabindex="-1"></a>save_dual_annot_heatmap(</span>
<span id="cb37-103"><a href="#cb37-103" aria-hidden="true" tabindex="-1"></a>    snp_df<span class="op">=</span>snp_df_ST38,</span>
<span id="cb37-104"><a href="#cb37-104" aria-hidden="true" tabindex="-1"></a>    ani_df<span class="op">=</span>ani_df_ST38,</span>
<span id="cb37-105"><a href="#cb37-105" aria-hidden="true" tabindex="-1"></a>    color_by<span class="op">=</span><span class="st">"snp"</span>,</span>
<span id="cb37-106"><a href="#cb37-106" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"ST38: SNP‐color &amp; ANI‐annot"</span>,</span>
<span id="cb37-107"><a href="#cb37-107" aria-hidden="true" tabindex="-1"></a>    rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb37-108"><a href="#cb37-108" aria-hidden="true" tabindex="-1"></a>    lower_legend<span class="op">=</span><span class="dv">0</span>,      <span class="co"># for the SNP colorbar</span></span>
<span id="cb37-109"><a href="#cb37-109" aria-hidden="true" tabindex="-1"></a>    upper_legend<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb37-110"><a href="#cb37-110" aria-hidden="true" tabindex="-1"></a>    output_png<span class="op">=</span><span class="st">"../imgs/ani_and_snp_heatmap_colored_by_snp_ST38.png"</span>,</span>
<span id="cb37-111"><a href="#cb37-111" aria-hidden="true" tabindex="-1"></a>    output_svg<span class="op">=</span><span class="st">"../imgs/ani_and_snp_heatmap_colored_by_snp_ST38.svg"</span></span>
<span id="cb37-112"><a href="#cb37-112" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/cell-16-output-1.png" width="1092" height="949" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Saved heatmap to ../imgs/ani_and_snp_heatmap_colored_by_ani_ST38.png and ../imgs/ani_and_snp_heatmap_colored_by_ani_ST38.svg</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/cell-16-output-3.png" width="1078" height="949" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Saved heatmap to ../imgs/ani_and_snp_heatmap_colored_by_snp_ST38.png and ../imgs/ani_and_snp_heatmap_colored_by_snp_ST38.svg</code></pre>
</div>
</div>
</section>
<section id="combined-ani-and-snp-values-annotated-with-metdata" class="level3">
<h3 class="anchored" data-anchor-id="combined-ani-and-snp-values-annotated-with-metdata">4.2 - Combined ANI and SNP values annotated with metdata</h3>
<div id="a9a935e8" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> save_dual_annot_heatmap_with_metadata(</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    snp_df,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    ani_df,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    metadata_df,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    metadata_col<span class="op">=</span><span class="st">"SiteCharacteristic"</span>,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    sample_id_col<span class="op">=</span><span class="st">"sample"</span>,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    color_map<span class="op">=</span><span class="va">None</span>,  <span class="co"># &lt;-- NEW optional argument</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    color_by<span class="op">=</span><span class="st">"ani"</span>,</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"Dual‐Annotated Heatmap with Metadata"</span>,</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>    output_png<span class="op">=</span><span class="st">"dual_heatmap_colored.png"</span>,</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>    output_svg<span class="op">=</span><span class="st">"dual_heatmap_colored.svg"</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a><span class="co">    Generates and saves a dual-annotated heatmap with colored axis labels.</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a><span class="co">    Accepts an optional dictionary to map characteristics to specific colors.</span></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1) Quick sanity checks</span></span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> snp_df.shape <span class="op">==</span> ani_df.shape, <span class="st">"Shapes of SNP and ANI dataframes must match"</span></span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> <span class="bu">all</span>(snp_df.index <span class="op">==</span> ani_df.index) <span class="kw">and</span> <span class="bu">all</span>(snp_df.columns <span class="op">==</span> ani_df.columns)</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> sample_id_col <span class="kw">in</span> metadata_df.columns, <span class="ss">f"'</span><span class="sc">{</span>sample_id_col<span class="sc">}</span><span class="ss">' not found in metadata"</span></span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> metadata_col <span class="kw">in</span> metadata_df.columns, <span class="ss">f"'</span><span class="sc">{</span>metadata_col<span class="sc">}</span><span class="ss">' not found in metadata"</span></span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 2) Prepare metadata and create a color map</span></span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>    meta_indexed <span class="op">=</span> metadata_df.set_index(sample_id_col)</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>    heatmap_samples <span class="op">=</span> snp_df.index</span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> color_map:</span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Use the user-provided color map</span></span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a>        site_color_map <span class="op">=</span> color_map</span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Check if all characteristics in the data have a color mapping</span></span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a>        all_sites_in_data <span class="op">=</span> meta_indexed.loc[heatmap_samples, metadata_col].unique()</span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a>        missing_keys <span class="op">=</span> [site <span class="cf">for</span> site <span class="kw">in</span> all_sites_in_data <span class="cf">if</span> site <span class="kw">not</span> <span class="kw">in</span> site_color_map]</span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> missing_keys:</span>
<span id="cb40-34"><a href="#cb40-34" aria-hidden="true" tabindex="-1"></a>            <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="ss">f"The provided color_map is missing colors for: </span><span class="sc">{</span><span class="st">', '</span><span class="sc">.</span>join(missing_keys)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb40-35"><a href="#cb40-35" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb40-36"><a href="#cb40-36" aria-hidden="true" tabindex="-1"></a>        <span class="co"># No map provided, generate one automatically</span></span>
<span id="cb40-37"><a href="#cb40-37" aria-hidden="true" tabindex="-1"></a>        unique_sites <span class="op">=</span> meta_indexed.loc[heatmap_samples, metadata_col].unique()</span>
<span id="cb40-38"><a href="#cb40-38" aria-hidden="true" tabindex="-1"></a>        palette <span class="op">=</span> sns.color_palette(<span class="st">"hls"</span>, <span class="bu">len</span>(unique_sites))</span>
<span id="cb40-39"><a href="#cb40-39" aria-hidden="true" tabindex="-1"></a>        site_color_map <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(unique_sites, palette))</span>
<span id="cb40-40"><a href="#cb40-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-41"><a href="#cb40-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 3) Pick the matrix that drives the color, mask, and set bounds</span></span>
<span id="cb40-42"><a href="#cb40-42" aria-hidden="true" tabindex="-1"></a>    cmap_df <span class="op">=</span> ani_df <span class="cf">if</span> color_by <span class="op">==</span> <span class="st">"ani"</span> <span class="cf">else</span> snp_df</span>
<span id="cb40-43"><a href="#cb40-43" aria-hidden="true" tabindex="-1"></a>    cbar_label <span class="op">=</span> <span class="st">"ANI (%)"</span> <span class="cf">if</span> color_by <span class="op">==</span> <span class="st">"ani"</span> <span class="cf">else</span> <span class="st">"SNP distance"</span></span>
<span id="cb40-44"><a href="#cb40-44" aria-hidden="true" tabindex="-1"></a>    cmap_color <span class="op">=</span> <span class="st">"coolwarm"</span> <span class="cf">if</span> color_by <span class="op">==</span> <span class="st">"ani"</span> <span class="cf">else</span> <span class="st">"RdBu_r"</span></span>
<span id="cb40-45"><a href="#cb40-45" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.triu(np.ones_like(cmap_df, dtype<span class="op">=</span><span class="bu">bool</span>))</span>
<span id="cb40-46"><a href="#cb40-46" aria-hidden="true" tabindex="-1"></a>    vmin <span class="op">=</span> np.nanmin(cmap_df.values[<span class="op">~</span>mask])</span>
<span id="cb40-47"><a href="#cb40-47" aria-hidden="true" tabindex="-1"></a>    vmax <span class="op">=</span> np.nanmax(cmap_df.values[<span class="op">~</span>mask])</span>
<span id="cb40-48"><a href="#cb40-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-49"><a href="#cb40-49" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 4) Set up the plot</span></span>
<span id="cb40-50"><a href="#cb40-50" aria-hidden="true" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">13</span>, <span class="dv">10</span>))</span>
<span id="cb40-51"><a href="#cb40-51" aria-hidden="true" tabindex="-1"></a>    sns.heatmap(</span>
<span id="cb40-52"><a href="#cb40-52" aria-hidden="true" tabindex="-1"></a>        cmap_df, mask<span class="op">=</span>mask, cmap<span class="op">=</span>cmap_color, vmin<span class="op">=</span>vmin, vmax<span class="op">=</span>vmax,</span>
<span id="cb40-53"><a href="#cb40-53" aria-hidden="true" tabindex="-1"></a>        annot<span class="op">=</span><span class="va">False</span>, cbar_kws<span class="op">=</span>{<span class="st">"label"</span>: cbar_label}, ax<span class="op">=</span>ax</span>
<span id="cb40-54"><a href="#cb40-54" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb40-55"><a href="#cb40-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-56"><a href="#cb40-56" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 5) Overlay both ANI and SNP text annotations</span></span>
<span id="cb40-57"><a href="#cb40-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(cmap_df.index)):</span>
<span id="cb40-58"><a href="#cb40-58" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(cmap_df.columns)):</span>
<span id="cb40-59"><a href="#cb40-59" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> mask[i, j]: <span class="cf">continue</span></span>
<span id="cb40-60"><a href="#cb40-60" aria-hidden="true" tabindex="-1"></a>            txt <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>ani_df<span class="sc">.</span>iloc[i, j]<span class="sc">:.1f}</span><span class="ss">%</span><span class="ch">\n</span><span class="sc">{</span><span class="bu">int</span>(snp_df.iloc[i, j])<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb40-61"><a href="#cb40-61" aria-hidden="true" tabindex="-1"></a>            ax.text(j <span class="op">+</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>, txt, ha<span class="op">=</span><span class="st">"center"</span>, va<span class="op">=</span><span class="st">"center"</span>, fontsize<span class="op">=</span><span class="dv">8</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb40-62"><a href="#cb40-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-63"><a href="#cb40-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 6) Apply colors to tick labels</span></span>
<span id="cb40-64"><a href="#cb40-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> tick_label <span class="kw">in</span> ax.get_yticklabels():</span>
<span id="cb40-65"><a href="#cb40-65" aria-hidden="true" tabindex="-1"></a>        tick_label.set_color(site_color_map[meta_indexed.loc[tick_label.get_text(), metadata_col]])</span>
<span id="cb40-66"><a href="#cb40-66" aria-hidden="true" tabindex="-1"></a>        tick_label.set_weight(<span class="st">'bold'</span>)</span>
<span id="cb40-67"><a href="#cb40-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> tick_label <span class="kw">in</span> ax.get_xticklabels():</span>
<span id="cb40-68"><a href="#cb40-68" aria-hidden="true" tabindex="-1"></a>        tick_label.set_color(site_color_map[meta_indexed.loc[tick_label.get_text(), metadata_col]])</span>
<span id="cb40-69"><a href="#cb40-69" aria-hidden="true" tabindex="-1"></a>        tick_label.set_weight(<span class="st">'bold'</span>)</span>
<span id="cb40-70"><a href="#cb40-70" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb40-71"><a href="#cb40-71" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 7) Add a custom legend for the site colors</span></span>
<span id="cb40-72"><a href="#cb40-72" aria-hidden="true" tabindex="-1"></a>    legend_patches <span class="op">=</span> [mpatches.Patch(color<span class="op">=</span>color, label<span class="op">=</span>site) <span class="cf">for</span> site, color <span class="kw">in</span> site_color_map.items()]</span>
<span id="cb40-73"><a href="#cb40-73" aria-hidden="true" tabindex="-1"></a>    ax.legend(handles<span class="op">=</span>legend_patches, title<span class="op">=</span>metadata_col.replace(<span class="st">'_'</span>, <span class="st">' '</span>).title(),</span>
<span id="cb40-74"><a href="#cb40-74" aria-hidden="true" tabindex="-1"></a>              bbox_to_anchor<span class="op">=</span>(<span class="fl">0.8</span>, <span class="dv">1</span>), loc<span class="op">=</span><span class="st">'upper center'</span>, borderaxespad<span class="op">=</span><span class="fl">0.</span>)</span>
<span id="cb40-75"><a href="#cb40-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-76"><a href="#cb40-76" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 8) Final touches</span></span>
<span id="cb40-77"><a href="#cb40-77" aria-hidden="true" tabindex="-1"></a>    ax.set_title(title, fontsize<span class="op">=</span><span class="dv">16</span>, pad<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb40-78"><a href="#cb40-78" aria-hidden="true" tabindex="-1"></a>    plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>, ha<span class="op">=</span><span class="st">"right"</span>)</span>
<span id="cb40-79"><a href="#cb40-79" aria-hidden="true" tabindex="-1"></a>    plt.yticks(rotation<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb40-80"><a href="#cb40-80" aria-hidden="true" tabindex="-1"></a>    fig.tight_layout(rect<span class="op">=</span>[<span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.9</span>, <span class="dv">1</span>])</span>
<span id="cb40-81"><a href="#cb40-81" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb40-82"><a href="#cb40-82" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 9) Save and show</span></span>
<span id="cb40-83"><a href="#cb40-83" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_png, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb40-84"><a href="#cb40-84" aria-hidden="true" tabindex="-1"></a>    plt.savefig(output_svg, <span class="bu">format</span><span class="op">=</span><span class="st">"svg"</span>)</span>
<span id="cb40-85"><a href="#cb40-85" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb40-86"><a href="#cb40-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-87"><a href="#cb40-87" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Heatmap saved to </span><span class="sc">{</span>output_png<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>output_svg<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now plot the heatmap</p>
<div id="4eee07da" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Your metadata</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>metadata_df <span class="op">=</span> pd.read_csv(<span class="st">"../data/site_characteristics.csv"</span>, sep<span class="op">=</span><span class="st">","</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>metadata_df <span class="op">=</span> pd.read_csv(<span class="st">"/Users/richard.goodman/Library/CloudStorage/OneDrive-LSTM/Github/trycycle-ESBL-E-jakarta/data/site_characteristics.csv"</span>, sep<span class="op">=</span><span class="st">","</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>metadata_df[<span class="st">'sample_name'</span>] <span class="op">=</span> metadata_df.index</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>metadata_df <span class="op">=</span> pd.DataFrame(metadata_df, columns<span class="op">=</span>[<span class="st">"sample_name"</span>, <span class="st">"SiteCharacteristic"</span>])</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>my_color_map <span class="op">=</span> {</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"01_Hospital"</span>: <span class="st">"#336699"</span>,</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"02_Community"</span>: <span class="st">"#85a3c2"</span>,</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>  <span class="st">"03_Market"</span>: <span class="st">"#9900ff"</span>,</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"04_Slaughterhouse"</span>: <span class="st">"#c57fff"</span>,</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"05_Upstream"</span>: <span class="st">"#339900"</span>,</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"07_Downstream"</span>: <span class="st">"#7fbf40"</span>,</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"08_Hospital wastewater inlet"</span>: <span class="st">"#b2d88c"</span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Now, call the updated function ---</span></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>save_dual_annot_heatmap_with_metadata(</span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>    snp_df<span class="op">=</span>snp_df_ST38,</span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>    ani_df<span class="op">=</span>ani_df_ST38,</span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a>    metadata_df<span class="op">=</span>metadata_df,</span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a>    metadata_col<span class="op">=</span><span class="st">"SiteCharacteristic"</span>, <span class="co"># The column to color by</span></span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>    sample_id_col<span class="op">=</span><span class="st">"sample_name"</span>,</span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>    color_map<span class="op">=</span>my_color_map,           <span class="co"># The column with sample IDs</span></span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a>    output_png<span class="op">=</span><span class="st">"../imgs/dual_heatmap_colored_ST38.png"</span>,</span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a>    output_svg<span class="op">=</span><span class="st">"../imgs/dual_heatmap_colored_ST38.svg"</span></span>
<span id="cb41-35"><a href="#cb41-35" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Fig_3A_ANI_SNP_matrix_for_specific_STs_files/figure-html/cell-18-output-1.png" width="1059" height="951" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Heatmap saved to ../imgs/dual_heatmap_colored_ST38.png and ../imgs/dual_heatmap_colored_ST38.svg</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>